### DISCUSSION TITLE
Microcks error: $objectToArray requires a document input, found: missing

### DISCUSSION THREAD

**Discussion by etumwizere21** [microcks/microcks#1370] [Community]
Hi everyone,
I'm trying to use Microcks create a mock server from my asyncapi definition file but whenever I upload the file I get this type of error and it does not create mocks though it shows the operations. I'm not sure if I'm missing anything. any thought?

failed to handle {"aggregate" : "service", "pipeline" : [ { "$project" : { "labels" : { "$objectToArray" : "$metadata.labels" } } }, { "$unwind" : "$labels" }, { "$sort" : { "labels.v" : -1 } }, { "$group" : { "_id" : "$labels.k", "values" : { "$addToSet" : "$labels.v" }, "key" : { "$first" : "$labels.k" } } } ], "cursor" : {}, "allowDiskUse" : false, "$db" : "test"}

de.bwaldvogel.mongo.backend.aggregation.Aggregation$PlanExecutorError: [Error 40390] PlanExecutor error during aggregation :: caused by :: $objectToArray requires a document input, found: missing
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.computeResult(Aggregation.java:242)
	at de.bwaldvogel.mongo.backend.AbstractMongoDatabase.commandAggregate(AbstractMongoDatabase.java:623)
	at de.bwaldvogel.mongo.backend.AbstractMongoDatabase.handleCommand(AbstractMongoDatabase.java:132)
	at de.bwaldvogel.mongo.backend.AbstractMongoBackend.handleCommand(AbstractMongoBackend.java:340)
	at de.bwaldvogel.mongo.backend.AbstractMongoBackend.handleMessage(AbstractMongoBackend.java:413)
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.handleMessage(MongoDatabaseHandler.java:73)
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.channelRead0(MongoDatabaseHandler.java:62)
	at de.bwaldvogel.mongo.wire.MongoDatabaseHandler.channelRead0(MongoDatabaseHandler.java:27)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base@17.0.12/java.lang.Thread.run(Thread.java:840)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:915)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:890)
Caused by: de.bwaldvogel.mongo.exception.MongoServerError: [Error 40390] $objectToArray requires a document input, found: missing
	at de.bwaldvogel.mongo.backend.aggregation.Expression$50.apply(Expression.java:1033)
	at de.bwaldvogel.mongo.backend.aggregation.Expression$50.apply(Expression.java:1028)
	at de.bwaldvogel.mongo.backend.aggregation.Expression.apply(Expression.java:1746)
	at de.bwaldvogel.mongo.backend.aggregation.Expression.evaluateDocumentExpression(Expression.java:1833)
	at de.bwaldvogel.mongo.backend.aggregation.Expression.evaluate(Expression.java:1811)
	at de.bwaldvogel.mongo.backend.aggregation.Expression.evaluateDocument(Expression.java:1752)
	at de.bwaldvogel.mongo.backend.aggregation.stage.ProjectStage.projectDocument(ProjectStage.java:82)
	at java.base@17.0.12/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base@17.0.12/java.util.ArrayList$Itr.forEachRemaining(ArrayList.java:1003)
	at java.base@17.0.12/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base@17.0.12/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base@17.0.12/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base@17.0.12/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base@17.0.12/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base@17.0.12/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@17.0.12/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at de.bwaldvogel.mongo.backend.aggregation.stage.GroupStage.apply(GroupStage.java:43)
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.runStagesAsStream(Aggregation.java:204)
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.runStages(Aggregation.java:196)
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.runStages(Aggregation.java:192)
	at de.bwaldvogel.mongo.backend.aggregation.Aggregation.computeResult(Aggregation.java:235)
	... 30 common frames omitted

here is our asyncapi definition attached

[message_new_v3.zip](https://github.com/user-attachments/files/17310403/message_new_v3.zip)

**Comment by lbroudoux** [comment]
Hello @etumwizere21 

From what I see you're using the `Uber` distribution of Microcks. When using this distribution, the stacktrace above is "normal" and due to the embedded striped-down version of MongoDB we're using. For some reason, it fails to aggregate information where no label are present (that's the case where you're starting an instance). Once a label has been put on an API or Service, this error disappears from the logs. 

Now it looks like this is not related at all to the import error you encounter. I'll have to take a look at your AsyncAPI file definition to see what's going wrong.

Final thought: using the uber image for doing asynchronous mocks is not the easiest path to go with when starting with Microcks. I would rather recommend using the Docker compose development mode (see https://microcks.io/documentation/guides/installation/docker-compose/#development-mode) for easier startup.

**Reply by etumwizere21** [reply]
Hi Laurent,
Thank you for getting back to me.  yes you are right I'm using microcks-uber I'm not able to understand the label you are talking about, can you please clarify the kind of label I'm missing because i did not see any field to fill in the label value while I was uploading the file?
actualy before going to uber distribution, I had tried the docker compose one but whenever I was trying to access locally(using localhost) the page couldn't come sucessfully and in the logs I could see the logs related to mongoDB container like the microcks container was not able to communicate with the mongoDB container, but I have have limited knowledge on this may be my understand was wrong.

I was reading the documentation from the above shared link and it says "docker compose -f docker-compose.yml -f docker-compose-async-addon.yml up -d" command will enable ASYNC API feature by enabling zookeeper kafka. we want to use MQTT for our asyncapi. don't you support it? 


kindly look at our AsyncAPI file definition an let me know if there is anything wrong or missing to have it pass.
thank you for your support

**Reply by lbroudoux** [reply]
I'm talking about those labels: https://microcks.io/documentation/guides/administration/organizing-repository/#1-applying-labels

I'll try to have a look at your file later on this evening.

**Reply by etumwizere21** [reply]
Hi Laurent,
were you able to look at the file? kindly let me know if you were able to learn the cause of the issue we are facing trying to use microcks.
Thank you!

**Reply by lbroudoux** [reply]
Hi @etumwizere21 

I have a look at your AsyncAPI file. The issue is that we expect to have an `operation. my operation.messages` list defined, pointing to a subset of your channel messages.

Your AsyncAPI file should then be chnaged that way:

```yaml
operations:
  receiveTelemetryInfo:
    action: receive
    channel:
      $ref: '#/channels/telemetry'
    summary: Receive telemetry information from a vehicle
    traits:
      - $ref: '#/components/operationTraits/mqtt'
    messages:
      - $ref: '#/channels/telemetry/messages/telemetry'
   
  receiveMobilityInfo:
    action: receive
    channel:
      $ref: '#/channels/mobility'
    summary: Receive mobility information from a vehicle
    traits:
      - $ref: '#/components/operationTraits/mqtt'
    messages:
      - $ref: '#/channels/mobility/messages/mobility'
    
  receiveWcaInfo:
    action: receive
    channel:
      $ref: '#/channels/wca'
    summary: Receive WCA information from a vehicle
    traits:
      - $ref: '#/components/operationTraits/mqtt'
    messages:
      - $ref: '#/channels/wca/messages/wcas'
```

After adding this information, import in Microcks works successfully.

**Reply by etumwizere21** [reply]
Hi Laurent,
Thank you for your recommendation! after adding it I can see a change in the operations mocks section. However it still shows that the mocks are disabled any advise on that? attached is the screenshot after I uploaded the file.
![microcks](https://github.com/user-attachments/assets/e123f9bc-c9f3-4d4f-bfb4-24a82c519994)

**Reply by lbroudoux** [reply]
`Mocks are disabled` message appears in the UI when the `async-enabled` flag is set to false in the `features.properties` file. If you're using the docker-compose with async addon installation method, this shouldn't be the case. Have you checked https://microcks.io/blog/async-features-with-docker-compose/ ?

**Reply by etumwizere21** [reply]
I have tried using docker-compose as it is documented, but when I type localhost:8080 in the browser to access the microcks UI page, it can't even load yet the container has started. attached are some of the screenshot for running containers, some of the logs for microcks, and microcks-async-minion
![running containers](https://github.com/user-attachments/assets/33141d6b-4118-4e11-b471-a83ba25de2c6)
![microcks-container-logs](https://github.com/user-attachments/assets/d38dde2e-df57-40ca-81a9-d7923e767ffd)
![microcks-async-minion-logs4](https://github.com/user-attachments/assets/21f87762-2e7b-4c39-a86a-b21f9eed462a)
![microcks-async-minion-logs3](https://github.com/user-attachments/assets/069c191f-da06-4c54-9263-cb1f6bab6f36)
![microcks-async-minion-logs2](https://github.com/user-attachments/assets/7afe37c0-4bf9-4664-93d8-66a022644c20)
![microcks-async-minion-logs1](https://github.com/user-attachments/assets/0be72fff-b48f-4141-8903-af8ca7c4119c)
![error-page](https://github.com/user-attachments/assets/51571f62-a07b-47a9-8f1b-8d1e6e192198)

Kindly advise me how to go about it.
another question, in my asyncapi I'm using MQTT and I see here the kafka container is the one running how will mqtt will come into play?

is it possible to have call session with you? if it is possible let me know what it requires.

Thank you

**Reply by lbroudoux** [reply]
Hello,

Everything seems correctly started and bound... It's weird `localhost:8080` brings nothing in your browser. Have you tried `http://127.0.0.1:8080` though? It looks like a low-level docker network issue. Is it normal you have to `sudo` all the docker commands? (sorry if it's a dumb question, I'm not used to Ubuntu).

Regarding MQTT, we don't provide an MQTT as such, so you have to use/start one on your own. You used to have https://github.com/microcks/microcks/blob/master/minions/async/artemis.sh for test purpose but I'm not sure what is it's current status. Once you have a MQTT broker running at hand, you then have to edit the `config/application.properties` under the `install/docker-compose` folder to tell Microcks how to connect it.

You first have to add `MQTT` to the list of supported bindings and then add the connection properties like below:

```properties
%docker-compose.minion.supported-bindings=KAFKA,WS,MQTT

%docker-compose.mqtt.server=localhost:1883
%docker-compose.mqtt.username=microcks
%docker-compose.mqtt.password=microcks
```

Hope it helps,

**Reply by etumwizere21** [reply]
Hi Laurent,
yeah I'm using sudo because I haven't added my user to the docker group to have non-root users permission.by default docker command requires elevated permissions.

I was able to get the page up and after uploading the asyncapi the thing of mocks are disabled disappeared. now the challenge I have is in testing. I'm not sure if I'm doing it wrong but I'm not getting the result. attached are the screenshot of what I'm doing and what I'm getting. if I'm doing wrong or putting wrong data please advise what to be done. Thank you

![microcks_apis service dashboard](https://github.com/user-attachments/assets/d09d4022-33b1-4530-88e3-601b0a74021d)
![newTestPageWithFilledData](https://github.com/user-attachments/assets/96193263-0c46-4d52-b584-d0ecd620849d)
![afterClickingLaunchTestButton](https://github.com/user-attachments/assets/9a5a9ef0-832b-426d-af68-d6ada4f95b64)

 I tried using kafka-console but I can't see the sample message from the asyncapi example. attached is the kafka-console screenshot.
![kafka console](https://github.com/user-attachments/assets/d2779f69-5ca8-42cd-b882-61ff89ef386f)

**Reply by lbroudoux** [reply]
Hi @etumwizere21,

I had another look at your AsyncAPI file, and it appears that it misses information to get mocking working.

Your channel definition is suing a parameter like that:

```yaml
  telemetry:
    address: '{vehicle_guid}/telemetry'
    messages:
      telemetry:
        $ref: '#/components/messages/Telemetry'
    description: AutoDrive asset telemetry information
    parameters:
      vehicle_guid:
        $ref: '#/components/parameters/vehicle_guid'
```

but your parameter definition does have a location definition on line 509. You should update it to something like:

```yaml
  parameters:
    vehicle_guid:
      description: The vehicle guid.
      location: $message.payload#/header/vehId_num
```

Otherwise, Microcks doesn't know how to retrieve values for this parameter and tries pushing messages to a Kafka topic named `TerraLinkConnectDataFlowAPI-0.0.1-{vehicle_guid}-mobility` and because `{}` are invalid characters in Kafka, this fails.

After applying the fix on the parameter, I can now read the messages on local Kafka topic:

```sh
$ kcat -b localhost:9092 -t TerraLinkConnectDataFlowAPI-0.0.1-0-mobility
--- OUTPUT ---
% Auto-selecting Consumer mode (use -P or -C to override)
{"header":{"absolute_time_sec":0,"absolute_time_nsec":0,"relative_time_sec":0,"relative_time_nsec":0,"vehId_make":"string","vehId_model":"string","vehId_num":0,"message_counter":"string"},"trajectory":[{"x":0,"y":0,"bearing":0,"speed_m_p_sec":0,"conditions":"STOP","time":"string"}],"autonomy_mode":"STOPPED","autonomy_status":"string","trailer_status":"HITCHED","brake_percent":0,"throttle_percent":0,"regen_brake_percent":0,"steering_wheel_angle_rad":0,"abs_pos":{"lat_long_deg":{"latitude":0,"longitude":0},"elevation_m":0,"ellipsoid":"string","utm_pos_northing_m":0,"utm_pos_easting_m":0,"utm_zone_number_m":0,"bearing_rad":0,"absolute_position_stddev_m":0},"mission_id":"string","mission_status":"MISSION_STATUS_UNKNOWN","task_id":["string"],"task_status":["TASK_STATUS_UNKNOWN"],"task_time_secs":[0],"speed":0}
```

**Reply by etumwizere21** [reply]
if I understood you well you are saying that I miss
location: $message.payload#/header/vehId_num
right?

**Reply by lbroudoux** [reply]
Yes exactly!

**Reply by etumwizere21** [reply]
Hi Laurent,
Now I have the async mock server with kafka working. Thank you so much for your support! 
I want to do the POC for async mock server with MQTT since it would be our first preference to use.

referring to your above reply saying: 
" 
Regarding MQTT, we don't provide an MQTT as such, so you have to use/start one on your own. You used to have https://github.com/microcks/microcks/blob/master/minions/async/artemis.sh for test purpose but I'm not sure what is it's current status. Once you have a MQTT broker running at hand, you then have to edit the config/application.properties under the install/docker-compose folder to tell Microcks how to connect it.

You first have to add MQTT to the list of supported bindings and then add the connection properties like below:
%docker-compose.minion.supported-bindings=KAFKA,WS,MQTT

%docker-compose.mqtt.server=localhost:1883
%docker-compose.mqtt.username=microcks
%docker-compose.mqtt.password=microcks
"
Attached are the screenshots of what I tried by follow your Instructions but I'm unable to read the topic message.
Please advise me what I'm missing, or I misunderstood.
Thank you

![application properties](https://github.com/user-attachments/assets/36dae492-d8a2-4d7d-b9ee-d0f1e3d7cbd4)
![running-containers](https://github.com/user-attachments/assets/8a3f22f4-cc5b-467d-87a4-914d1b9b3c3d)
![Microcks-UI](https://github.com/user-attachments/assets/80a8d47e-491a-45ca-a3a3-9bed0831c60b)
![MQTT_sub_topic](https://github.com/user-attachments/assets/a2dc842e-2e61-4889-8ea3-6591134ceebe)
![sub_topic with provided-endpoint-from-Microcks-UI](https://github.com/user-attachments/assets/226032e0-dc69-42b2-ba0e-0f6b48f80c89)

**Reply by lbroudoux** [reply]
Hey @etumwizere21 
There are few things to change and check here.
1) You have to declare a MQTT bindings on your AsyncAPI file so that Microcks actually knows it has to connect to a broker and send messages. This can be done applying a binding like in https://github.com/microcks/microcks/blob/master/webapp/src/test/resources/io/github/microcks/util/asyncapi/api-maintenance.async-api-spec.yaml#L17 Beware that the sample if for AsyncAPI 2 and it has to be adapted if using AsyncAPI 3. The reference is here: https://github.com/asyncapi/bindings/tree/master/mqtt . Once this is done, you should see an additional `MQTT` bindings appear in the service details in Microcks UI.
2) Please check that https://github.com/microcks/microcks/blob/master/minions/async/artemis.sh is actually working on Linux as it is now broken for MacOS (we have to move to another image).
3) From the microcks-async-minion container perspective, the MQTT server is probably not `%docker-compose.mqtt.server=localhost:1883` as you're using another container. You might try: to use the MQTT container alias if you started this container within the same docker-compose file / docker network or try `host.docker.internal` to reach out your host network.

**Reply by etumwizere21** [reply]
Hi Laurent,
I rely appreciate your support! Thank you so much
I think I have tried to follow your instruction as the attached screenshot shows unless I misunderstood it. 
![serverConfig](https://github.com/user-attachments/assets/7ca5448f-c9d7-4570-85c4-096b90001d9c)
![message binding](https://github.com/user-attachments/assets/dde2776f-f59b-434e-849a-2f072b5735ed)
![application properties](https://github.com/user-attachments/assets/eb770e32-39b2-4758-9c9d-4ebad435f0f3)
![running containers](https://github.com/user-attachments/assets/81e63d88-b610-421d-b4d2-23b80406d8c1)
![microcks ui](https://github.com/user-attachments/assets/990fecba-5f0a-4c41-a26f-22dc86a2681c)

I think the problem I have now is with testing it. like subscribing to see the messages. below is what I was trying but doesn't work. please guide me on how you test it after having the asycaip uploaded.
![testing with mosquitto container](https://github.com/user-attachments/assets/04440a17-6986-4266-b718-7ef5a6fc099c)
![mqtt explorer](https://github.com/user-attachments/assets/32f8d6af-f887-4125-bca1-de9d76004cf8)

**Comment by lbroudoux** [comment]
Hello,
`my-mqtt-broker.apps.try.microcks.io` is just a placeholder we put as example in the `features.properties` configuration file. In your case, I think that `localhost:1883` would be a better option if your MQTT Explorer application is running as a native application on your host network.

Have you checked the async-minion container logs? Is the connection to Artemis from the minion is ok? Can you see in the logs that it is publishing mock messages on the queues?

**Reply by etumwizere21** [reply]
Hi Laurent,
Looking at minion logs, it looks like there are connection issues. in the application.properties I tried both %docker-compose.mqtt.server=localhost:1883 and docker-compose.mqtt.server=async-artemis-1 (async-artemis-1 being alias) but none solves the connection issues. below are the screenshot.

With "%docker-compose.mqtt.server=localhost:1883"
=======================================
Microcks-UI: 
![Microcks UI](https://github.com/user-attachments/assets/4aca2c37-b217-43f9-b011-a1928b3a13b7)

minion Running ContainerID:
![running containers](https://github.com/user-attachments/assets/affe3f20-bad0-4e53-8318-e087c29ab4b1)

Minion logs:
![minion logs1](https://github.com/user-attachments/assets/25185db4-152a-46e8-a4b6-8d75b2719ca2)
![minio lods 2](https://github.com/user-attachments/assets/fd57f57b-0aeb-454c-961e-50e17b99b76a)
 
With "docker-compose.mqtt.server=async-artemis-1"
=======================================
MQTT Running container Alias:
![MQTT Running contanainer alias](https://github.com/user-attachments/assets/cb101b25-7a51-42ba-b9da-6e8098542d5f)

Minion running containerID:
![minion containerID](https://github.com/user-attachments/assets/e4b2bd87-9f39-4644-9519-aa1fd64caa20)

Minion error logs:
![minion error logs for MQTT alias](https://github.com/user-attachments/assets/6f7def33-b4e9-4eb5-86b1-01aa1947fc71)
![minion error logs2](https://github.com/user-attachments/assets/df9a744e-9c05-4485-99f7-c0b93ab521e3)

**Reply by lbroudoux** [reply]
Are you sure that the `async-artemis-1` is actually attached to the same docker network? It is the case when you start all the services from the same docker-compose file but not if you have started it outside... And have you tried with `host.docker.internal:1883`?

**Reply by etumwizere21** [reply]
yes I have tried both. attached are screenhorts for host.docker.internal:1883  please let me know if we can have a working session with you anytime. I'm happy to meet what it requires to have a session with you and have this closed if at your end it is working as you expected.
![app properties](https://github.com/user-attachments/assets/ae37783d-cf2e-4f67-ad4c-931489fef8bb)
![running containerss](https://github.com/user-attachments/assets/4f1dd6e8-e466-4e8d-a103-d703da8d4073)
![log11](https://github.com/user-attachments/assets/28b1792a-b49b-45d5-a10b-478b269e7409)
![log12](https://github.com/user-attachments/assets/b266daba-8c23-42e7-bbc3-fda5342615f9)
![log13](https://github.com/user-attachments/assets/a6d0c26c-d37f-48c4-9a55-7a1f3f078453)

**Reply by etumwizere21** [reply]
Hi Laurent,
did you get a chance to look at the above ? any feedback?
